code/modules/atmospherics/machinery/components/components_base.dm

```diff
@@ -95,6 +96,15 @@
 	var/i = parents.Find(reference)
 	reference.other_airs -= airs[i]
 	reference.other_atmosmch -= src
+	/**
+	 *  We explicitly qdel pipeline when this particular pipeline
+	 *  is projected to have no member and cause GC problems.
+	 *  We have to do this because components don't qdel pipelines
+	 *  while pipes must and will happily wreck and rebuild everything again
+	 *  every time they are qdeleted.
+	 */
+	if(!(reference.other_atmosmch.len || reference.members.len || QDESTROYING(reference)))
```

Comment:
`if(already deleting) delete again` doesn't make much sense to me.

Reply:
Unless there's a horrible invisible code monster out there, `QDESTROYING` should return `TRUE` when the thing is already `Destroy()`ing or GCed, or `FALSE` otherwise, and that will stop the `qdel()`. There is a `!` operator overarching three expressions; Any of these not returning `FALSE` will stop the `qdel()`.

On a side note, we can go fast and loose with `QDEL_NULL(parents[i])` to unconnected nodes at `nullifyNode` which is going to be probably right and fast but not very safe nor fitting to the proc name. I won't take the risk and trust every code touching `parents[i]` or `nodes[i]`... but maybe Ghilker will?

Reply:
If `reference` is referenced by the `parents` variable _and_ it is being qdeleted, then it's going to hard delete. This would indicate an issue that needs to be addressed on the object's deletion, either through the Destroy proc or through the component registering to its destruction.

Reply:
I missed that negation was outside the parens, yeah this makes more sense.

@Rohesie  It's nulled right after the call ? I guess could be moved before qdel call.

Reply:
My point is that if this condition is ever true, then it should be screaming error. It should be notifying when it's happening so it's fixed:
```suggestion
	if(!length(reference.other_atmosmch) && !length(reference.members))
		if(QDESTROYING(reference))
			parents[i] = null
			CRASH("nullifyPipenet() called on qdeleting [reference] indexed on parents\[[i]\]")
```

Reply:
I was thinking about qdel called upon pipenets which only has components as members, (think directly connected components: the resulting pipenet won't have more than two members) for it was the only way the `QDESTROYING` was needed: at the last components' `nullifyPipenet` that gets called in the `other_atmosmch` iteration of said pipenets' `Destroy()`, the pipenet's both lists are empty (`members` list don't get cleaned at pipenet `Destroy()`) and the only way it could know if it needs to call `qdel(reference)` ends up being `QDESTROYING`. But after reviewing how `qdel()` is called upon a pipenet, I now think such cases shouldn't happen since `merge` `Cut()`s both lists before qdel, pipe `destroy_network()` qdel obviously has pipe member(s), and a new qdel here happens after clearing the last ref.

Therefore, `QDESTROYING` here returning `TRUE` means a serious recordkeeping error (or my mind's error? who knows what could be lying deep in the atmos code?) in `other_atmosmch`, `members` and/or `parents`, so, yeah, doing a `CRASH` instead is a good solution.